File: auntieos/web/composeApp/src/commonTest/kotlin/com/tribetails/auntieos/web/data/ChecklistBankTest.kt
package com.tribetails.auntieos.web.data

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ChecklistBankTest {

    @Test fun decodesItems() {
        val json = """{"items":[
            {"id":"a","text":"Fresh water","scope":"PER_PET"},
            {"id":"b","text":"Home secured","scope":"PER_VISIT"}
        ]}"""
        val out = decodeChecklistBank(json)
        assertEquals(2, out.size)
        assertEquals("Fresh water", out[0].text)
        assertEquals("PER_VISIT", out[1].scope)
    }

    @Test fun decode_missingItems_isEmpty() {
        assertEquals(emptyList(), decodeChecklistBank("{}"))
    }

    @Test fun bankItemsNotInChecklist_hidesAlreadyAddedByText() {
        val bank = listOf(
            ChecklistBankItem("a", "Fresh water", "PER_PET"),
            ChecklistBankItem("b", "Walk", "PER_PET"),
        )
        val current = listOf(ChecklistItem(key = "k1", text = "fresh water", scope = "PER_PET"))
        val out = bankItemsNotInChecklist(bank, current)
        assertEquals(listOf("Walk"), out.map { it.text })
    }

    @Test fun bankItemsNotInChecklist_keepsAllWhenNoOverlap() {
        val bank = listOf(ChecklistBankItem("a", "Fresh water", "PER_PET"))
        assertEquals(1, bankItemsNotInChecklist(bank, emptyList()).size)
    }

    @Test fun checklistItemFromBank_carriesTextScopeOrderKey() {
        val item = checklistItemFromBank(ChecklistBankItem("a", "Walk", "PER_VISIT"), key = "k9", order = 3)
        assertEquals("Walk", item.text)
        assertEquals("PER_VISIT", item.scope)
        assertEquals("k9", item.key)
        assertEquals(3, item.order)
    }

    @Test fun checklistItemContainsText_isCaseAndSpaceInsensitive() {
        val items = listOf(ChecklistItem(key = "k", text = "  Fresh Water  ", scope = "PER_PET"))
        assertTrue(checklistHasText(items, "fresh water"))
        assertFalse(checklistHasText(items, "walk"))
    }
}

File: auntieos-admin/src/lib/checklistBank.test.ts
import { describe, expect, it } from 'vitest';
import { makeChecklistItem } from './kinTale/model';
import {
  bankItemsForScope,
  bankItemsNotInChecklist,
  checklistHasText,
  checklistItemFromBank,
  type ChecklistBankItem,
} from './checklistBank';

/** Ported case-for-case from the archive's `ChecklistBankTest.kt`. */

function bankItem(over: Partial<ChecklistBankItem> = {}): ChecklistBankItem {
  return { id: 'a', text: 'Fresh water', scope: 'PER_PET', ...over };
}

describe('checklistHasText', () => {
  it('matches ignoring case and surrounding space, so the bank never offers a near-duplicate', () => {
    const items = [makeChecklistItem({ key: 'i1', text: 'Fresh Water' })];
    expect(checklistHasText(items, '  fresh water ')).toBe(true);
    expect(checklistHasText(items, 'Walk')).toBe(false);
  });
});

describe('bankItemsNotInChecklist', () => {
  it('drops bank entries already present in the checklist', () => {
    const bank = [bankItem({ id: 'a', text: 'Fresh water' }), bankItem({ id: 'b', text: 'Walk' })];
    const items = [makeChecklistItem({ key: 'i1', text: 'fresh water' })];
    expect(bankItemsNotInChecklist(bank, items).map((b) => b.text)).toEqual(['Walk']);
  });

  it('returns the whole bank when the checklist is empty', () => {
    const bank = [bankItem({ id: 'a' }), bankItem({ id: 'b', text: 'Walk' })];
    expect(bankItemsNotInChecklist(bank, [])).toHaveLength(2);
  });
});

describe('bankItemsForScope', () => {
  it('offers a scope only the items saved for it, case-insensitively', () => {
    const bank = [
      bankItem({ id: 'a', text: 'Fresh water', scope: 'PER_PET' }),
      bankItem({ id: 'b', text: 'Home secured', scope: 'per_visit' }),
    ];
    expect(bankItemsForScope(bank, [], 'PER_VISIT').map((b) => b.text)).toEqual(['Home secured']);
    expect(bankItemsForScope(bank, [], 'PER_PET').map((b) => b.text)).toEqual(['Fresh water']);
  });

  it('excludes an item already in the checklist even when the scope matches', () => {
    const bank = [bankItem({ id: 'a', text: 'Fresh water', scope: 'PER_PET' })];
    const items = [makeChecklistItem({ key: 'i1', text: 'Fresh water', scope: 'PER_PET' })];
    expect(bankItemsForScope(bank, items, 'PER_PET')).toEqual([]);
  });
});

describe('checklistItemFromBank', () => {
  it('carries the text and scope onto a fresh checklist item at the given key/order', () => {
    const item = checklistItemFromBank(bankItem({ text: 'Walk', scope: 'PER_VISIT' }), 'item_9', 3);
    expect(item).toMatchObject({ key: 'item_9', text: 'Walk', scope: 'PER_VISIT', order: 3 });
    // Everything else comes from the model default: a bank item carries no
    // conditions, so the new row is unconditionally visible.
    expect(item.conditions).toEqual([]);
    expect(item.required).toBe(false);
  });

  it('normalizes an unrecognized scope to PER_PET rather than persisting it', () => {
    const item = checklistItemFromBank(bankItem({ scope: 'SOMETHING_ELSE' }), 'item_1', 0);
    expect(item.scope).toBe('PER_PET');
  });
});
